File: testes/fib.java
import java.util.Scanner;
public class fib{
	private static long[] fib_cache;

	public static void main (String[] args){
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		fib_cache = new long[n + 1];
		System.out.println(fibonacci(n));
		sc.close();
	}

	public static long fibonacci(int n){
		if(n <= 1){
			return n;
		}
		if(fib_cache[n] != 0){
			return fib_cache[n];
		}
		long fibAns = fibonacci(n-2) + fibonacci(n-1);
		fib_cache[n] = fibAns;
		return fibAns;
	}

}

File: testes/fib.cpp
#include <bits/stdc++.h>
#define ll long long
using namespace std;

const ll *fib_cache;

ll fib(int n){
	if( n <= 1 ){
		return n;
	}
	if(fib_cache[n] != NULL){
		return fib_cache[n];
	}
	ll fibAns = fib(n-2) + fib(n-1);
	return fibAns;
}

int main(void){
	int n = 0; 
	cin >> n;
	fib_cache = new ll[n+1];
	cout << fib(n) << '\n';
	return(0);
}
